add-dbs/addDbs.cpp

```diff
@@ -0,0 +1,253 @@
+            exit(-1);
+        }
+        std::ofstream ss(output_file.string(), std::ios::out);
+        if(!ss.good())
+        {
+            std::cerr << "Unable to open output file: " << output_file.string() << std::endl;
+            exit(-1);
+        }
+        for(auto arch_idx = 0; arch_idx < arches.size(); arch_idx++)
+        {
+            auto kinder = arches[arch_idx];
+            auto arch   = kinder.first;
+            auto num_cu = kinder.second;
+            std::stringstream query_ss;
+            query_ss << "select key, group_concat(solver || ':' || params, ';') as res "
+                        "from config inner join perf_db on config.id = perf_db.config "
+                        "where arch = '"
+                     << arch << "' and num_cu = " << num_cu << " group by key";
+            result_type configs;
+            rc = sqlite3_exec(ptr_sql,
+                              query_ss.str().c_str(),
+                              find_callback,
+                              static_cast<void*>(&configs),
+                              nullptr);
+            if(rc != SQLITE_OK)
+            {
+                std::cerr << "Unable to execute query: " << query_ss.str() << std::endl;
+                exit(-1);
+            }
+            if(arch_idx != 0)
+                ss << " else ";
+
+            // ss << "if(arch_cu == \"" << arch << "-" << num_cu << "\") \n {\n static const
+            // std::unordered_map<std::string, CacheItem> data \n\t {\n";
+            ss << "if(arch_cu == \"" << arch << "_" << num_cu << "\") \n {\n static const "
+                                                                 "std::unordered_map<std::string, "
+                                                                 "std::string> data \n\t {\n";
+            // handle the joining comma
+            // ss << "\t\t { \"" << configs[0]["key"] << "\", { " << 0 << ", \"" <<
+            // configs[0]["res"]
+            //    << "\"}}" << std::endl;
+            ss << "\t\t { \"" << configs[0]["key"] << "\", \"" << configs[0]["res"] << "\"}"
+               << std::endl;
+            for(auto cfg_idx = 1; cfg_idx < configs.size(); cfg_idx++)
+            {
+                auto config = configs[cfg_idx];
+                if(!config["res"].empty())
+                    // ss << "\t\t,{ \"" << config["key"] << "\", { " << cfg_idx << ", \"" <<
+                    // config["res"]
+                    //    << "\"}}" << std::endl;
+                    ss << "\t\t,{ \"" << config["key"] << "\", \"" << config["res"] << "\"}"
+                       << std::endl;
+            }
+            ss << "};\n return data; \n}" << std::endl;
+        }
+        // default case for invalid/unknown arch-cu pair
+        ss << "else {\n static const std::unordered_map<std::string, std::string> data \n\t {\n}; "
+              "\n return data; \n}"
+           << std::endl;
+    }
+    else // plaintext
```

Comment:
Is there any reason at all to use such an approach instead of addkernels with GetKernelSource? At least for text DB. I know that naming would be misleading unless changed, but that would be much simpler in terms of required modifications to other source files with no nonmarginal performance loss.

Reply:
That is indeed a sound approach, my concern was to implement something which I can use easily for both Find and Perf Dbs and avoid memory overhead. Since using the file would mean that once it would be placed in memory as a file and then parsed out to created an in memory map. This way we directly create the in-memory map.

Reply:
> This way we directly create the in-memory map.

Oh, no, please. This creates implicit dependence. Let's reuse existing machinery. Let's do not care about file contents here at all. The existing parser (or whatever reader) knows the format. It knows how to read it, how to write it, and how the cached representation is stored in memory.

Reply:
If you want to avoid memory overhead store a unique_ptr pointer to the file contents and move it as a return, that will delete the blob upon reading.

Reply:
@atamazov Please elaborate how this creates implicit dependency ? We __are__ reusing existing machinery for parsing the key and the values.  Why should we defer something to runtime, when we can do it at compile time.

Reply:
> If you want to avoid memory overhead store a unique_ptr pointer to the file contents and move it as a return, that will delete the blob upon reading.

The text file would still be embedded and thus taking space in the binary, perhaps I am missing something?

Reply:
The code fragment started from [here](https://github.com/ROCmSoftwarePlatform/MIOpen/pull/262/files/98704c45aa71084196fc993a429698fc4baa1d14#r436869526) implicitly depends on:
- Database format.
- Implementation of read-only cache (how it holds cached data in memory).

> We are reusing existing machinery for parsing the key and the values. 

Well, unfortunately this looks like copy-and-paste, which is IMO the "cheap reuse" that introduces deferred costs.

Reply:
> The text file would still be embedded and thus taking space in the binary, perhaps I am missing something?

You are correct. However if we hold it zipped, then the size will increase much less (~1/10th).

Reply:
> You are correct. However if we hold it zipped, then the size will increase much less (~1/10th).

So you would prefer an indirect method with the added complexity of zipping data instead of directly storing the data in the required format?

Reply:
> Well, unfortunately this looks like copy-and-paste, which is IMO the "cheap reuse" that introduces deferred costs.

I cringed at that use too, however, I could not use code from MIOpen itself since that creates a circular dependency in the MIOpen build process (MIOpen <=> embed) and there is no way to avoid it.

Reply:
> So you would prefer an indirect method with the added complexity of zipping data instead of directly storing the data in the required format?

Almost correct. Actually I would prefer removal of as much complexity as possible, and store the raw content of the file without zipping. In this case the size of the library (on disk) would be bigger, but the code should be simpler (i.e. less implicit dependencies)

Zipping is orthogonal feature that could be implemented separately. However I suspect that we will want to optimize library size anyway, so I am proposing zipping at build time/unzipping at runtime. The overhead is some additional time and memory required once during initialization. However I am expecting that this would be anyway "better" (i.e. needs less resources) than current implementation that reads the whole file from disk.

___Both "direct storing the data in the required format" method and "raw data blob" method___ occupy ~10X more space in the library than compressed blob. In addition, compressing would save build time.

And yes, the cost of these extras is added complexity, as you correctly noted.

Reply:
> Almost correct. Actually I would prefer removal of as much complexity as possible, and store the raw content of the file without zipping. In this case the size of the library (on disk) would be bigger (than compressed, but almost equal to static initialization), but the code should be simpler (i.e. less implicit dependencies)

I mentioned earlier, this would result in two copies of the file, once in the binary ( and therefore in memory) and once in memory when we parse (and unzip) it. This is not ideal because some of our customers are very sensitive to disk and memory usage. Also, it is unclear how embedding the file would be less complex than simply embedding the map, since the map would be reconstructed from the embedded file.

> Zipping is orthogonal feature that could be implemented separately. However I suspect that we will want to optimize library size anyway, so I am proposing zipping at build time/unzipping at runtime. The overhead is some additional time and memory required once during initialization. However I am expecting that this would be anyway "better" (i.e. needs less resources) than current implementation that reads the whole file from disk.

Again, this adds an additional subsequent step, to very little tangible advantage.

> _**Both "direct storing the data in the required format" method and "raw data blob" method**_ occupy ~10X more space in the library than compressed blob. In addition, compressing would save build time.

The overhead on build time is __only__ when we are building the package, not during development. The average developer does not need the files embedded in the binary and therefore can live happily without enabling the CMake variable.

Reply:
> I mentioned earlier, this would result in two copies of the file, once in the binary (and therefore in memory) and once in memory when we parse (and unzip) it.

This is complex question, easy to mistake.

The compiler is often able to statically (at compile time) initialize constant objects. When this happens, the object does not need to be constructed at runtime and occupy memory (except its initializer). Any access to the object would refer to the pre-constructed data in the loaded .rodata section. 

For example, blobs can be defined as
```C++
const char blob[] = "...";
```
and require no initialization (and no copy) at run time.

However, C++ is able to do that only if initialization is _sufficiently simple_. I am afraid that in the case of `std::unordered_map<std::string, std::string>`, only string literals will reside in .rodata, but the map will be constructed (and initialized) at run time.

> it is unclear how embedding the file would be less complex than simply embedding the map, since the map would be reconstructed from the embedded file.

Of course, this is not less complex from the point of view of a computer, but I do not think about it (unless optimization for speed is required). I would like to make it less complex for humans. Less implicit dependencies and less new code -> less complexity & easier maintenance. This is a value judgment, however.

> ...some of our customers are _very sensitive_ to disk and memory usage....
> ...\<compression\> adds an additional subsequent step, to _very little_ tangible advantage.

Absolutely agree with both ;)

## Comparison against current implementation (develop):

  | Current implementation | Static initialization | Blob | Compressed blob
-- | -- | -- | -- | --
Size on disk per GPU type, ~MB | 15 (extermnal file) | 15 | 15 | 1.5
Size of .rodata section per   GPU type, ~MB (this affects memory footprint) | 0 | 15 | 15 | 1.5
Extra time during C++ startup | No | Yes | No | No
Extra time during library init | Yes | No | Yes | Yes
Temporary buffer in RAM during   library init | No | No | No | Yes, ~15MB
New implicit dependencies | No | Yes | No | No
Additional app (embed) | No | Yes | No | No
Modification of addkernels   required | No | No | No | Yes (zipping)
Modification of library | No | More than others | Minimal | Minimal + unzipping

Reply:
### Size on dick, memory footprint etc

  | Current implementation | Static initialization | Blob | Compressed blob
-- | -- | -- | -- | --
Size on disk per GPU type, ~MB | 15 (extermnal file) | 15 | 15 | 1.5
Size of .rodata section per   GPU type, ~MB (this affects memory footprint) | 0 | 15 | 15 | 1.5
Extra time during C++ startup | No | Yes | No | No
Extra time during library init | Yes | No | Yes | Yes
Temporary buffer in RAM during   library init | No | No | No | Yes, ~15MB

### Complexity

  | Current implementation | Static initialization | Blob | Compressed blob
-- | -- | -- | -- | --
New implicit dependencies | No | Yes | No | No
Additional app (embed) | No | Yes | No | No
Modification of addkernels   required | No | No | No | Yes (zipping)
Modification of library | No | More than others | Minimal | Minimal + unzipping

Reply:
Thanks for the analysis, however, I am not sure I get your point.

Reply:
;) In short, the costs of construction of the const static map from huge initializer match the costs of construction of static map at run-time (except parsing the text buffer and, if used, unzipping).

I think that the closest approximation to the current implementation in terms of performance, memory footprint and amount of work, would be using compressed text.

Reply:
> ;) In short, the costs of construction of the const static map from huge initializer match the costs of construction of static map at run-time (except parsing the text buffer and, if used, unzipping).

The key is the except part, all of which is a persistent runtime cost, __every__ time MIOpen access it.

> I think that the closest approximation to the current implementation in terms of performance, memory footprint and amount of work, would be using compressed text.

You do realize that is additional complexity and avoidable given this PR.

Reply:
> The key is the except part, all of which is a persistent runtime cost, every time MIOpen accesses it.

Oh, no, no, initialization happens once per library instance. 

The "exept part" includes only pasring and unzipping. Parsing is really quick. Unzipping should be at least as fast as reading the file.

Other things -- constructing strings from string literals and construsting the map from those, -- happen identically in all four methods of initialization.

Reply:
How should I proceed ?

Reply:
WRT perf-db, I suppose you decided to embed it using the same mechanisms used for precompiled kernels. That is, somehow embed SQLite db? instantiate is at run-time and work with it as usual (almost). Here I have nothing special to say.

Regarding find-db, I would advise converting the text to a compressed blob using (modified) addkernels. At runtime, it is enough to unpack it into a temporary buffer once, and then initialize the already existing read-only system find-db cache from it (instead of reading file). This can be done by Vasily or me in a separate PR, if you wish.

Reply:
One comment about embedding sqlite databases. If SQLite is able to work with `const char[]` buffers instead of files, then we can re-use addkernels.

Also we can use zipping to reduce file size, however that would increase memory footprint. I tried zipping, compression ratio is ~5:1 for perf-db and ~2:1 (or worse? this is very approximate) for kernel db.